src/Http/Routing/src/IDynamicEndpointMetadata.cs

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    /// <summary>
+    /// A metadata interface that can be used to specify that the associated <see cref="Endpoint" />
+    /// will be dynamically replaced during matching. 
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// <see cref="IDynamicEndpointMetadata"/> and related derived interfaces signal to
+    /// <see cref="MatcherPolicy"/> implementations that an <see cref="Endpoint"/> has dynamic behavior
+    /// and thus cannot have its characteristics cached.
+    /// </para>
+    /// <para>
+    /// Using dynamic endpoints can be useful because the default matcher implementation does not
+    /// supply extensibility for how URLs are processed. Routing implementations that have dynamic
+    /// behavior can apply their dynamic logic after URL processing, by replacing a endpoints as
+    /// part of a <see cref="CandidateSet"/>.
+    /// </para>
+    /// </remarks>
+    public interface IDynamicEndpointMetadata
+    {
+        /// <summary>
+        /// Returns a value that indicates whether the associated endpoint has dynamic matching
+        /// behavior.
+        /// </summary>
+        bool IsDynamic { get; }
+    }
```

Comment:
I also considered another subclass of `Endpoint` and you can into trouble. You end up wanting to subclass `RouteEndpoint` instead.

Then the next problem you run into is that `Endpoint`s are **data** - but making a runtime decision about what to do always requires **services**. Putting your services into your data is always a mistake that will cause pain in the future.

From this logic, I ended up thinking of dynamic endpoints as a pattern instead of infrastructure.

Reply:
Imagine that we also used this same pattern to build something like:

```
routes.MapControllers();
routes.MapDynamicControllerRoute("{*path}", async (context, values) =>
{
    // Write your CMS here
   
    return someEndpoint;
});
```

The idea is that we give you a services that you can call to do something like MVC's action selection, and you can return the endpoint that you want (after doing your stuff). This is the pattern that I've seen used most often with `IRouter`.

Reply:
`IDynamicEndpointMetadata.IsDynamic => false;` :trollface: Is the presence of the interface not sufficient?